### PR TITLE
Limit maximum merged module name length

### PR DESCRIPTION
**Pull request by steveniemitz**
The module merger can end up generating module names that are so long they exceed the maximum file name length and can't be saved.  This limits them to a saner 150 (arbitrarily chosen) characters, plus an MD5 hash of all the target names to ensure they remain unique.

### CODE REVIEW

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (unkarjedy) [inline bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: Consider using `org.apache.commons.codec.digest.DigestUtils#md5Hex(java.lang.String)`

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (unkarjedy) [inline bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: We can use `org.apache.commons.io.FileSystem.getCurrent` + `org.apache.commons.io.FileSystem#getMaxFileNameLength` (it's 255 for all OS)

(we should remember to substract suffix length when calling `substring`)

**Comment by abrams27** [comment]
is it possible that names arent unique? the name is a path + suffix (names separated with `+` if im not mistaken), paths are unique in principle, so these names should also be unique

for bazel ofc xd

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (unkarjedy) [inline bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: I am maybe a little out of the BSP context, but can't we just call md5 on `ret` itself?

**Comment by steveniemitz** [comment]
> is it possible that names arent unique? the name is a path + suffix (names separated with `+` if im not mistaken), paths are unique in principle, so these names should also be unique
> 
> for bazel ofc xd

Sorry I'm not sure what your question is exactly?  The hashed version of the names should be just as unique as the full text, the chance of a hash collision here is incredibly small.

**Comment by abrams27** [comment]
so i was wondering if there is a case where the hash would be needed (is there a case where the module name is not unique), but I forgot that you are trimming the name now, so it is possible

**Comment by steveniemitz** [comment]
> so i was wondering if there is a case where the hash would be needed (is there a case where the module name is not unique), but I forgot that you are trimming the name now, so it is possible

ah, it's needed because the name is getting trimmed down yeah.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (jastice) [inline bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: Just wondering if we can make this a little smarter so as to not have 2 very similar module names, or is it a rare enough occurrence not to matter in practice?

Comment (steveniemitz) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: if you have any suggestions for making it smarter I'm all ears. :P

It really shouldn't be very common though, I only ran into this because we have a pretty weird edge case in our build system that causes many modules to all contain the same source root.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (steveniemitz) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: I mean at that point, does 150 vs 255 characters really matter?  It doesn't need to be exact here, once you're past like 100 the name is pretty meaningless.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (steveniemitz) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: I don't see why that wouldn't work, nor why it would be any better than what its doing now.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (unkarjedy) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: It's of cause not a major.
There is just no need in an extra constant.
Code readers and future maintainers will question: "what is this 150, why it is 150"?
Using `getMaxFileNameLength` will clearly show the intention and cause.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala
Comment (unkarjedy) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: > why it would be any better than what its doing now.

1. Ret is already calculated, no extra calculation is required.
2. Again, code maintainers will question "why using `targets.map(_.getDisplayName).mkString`, is it some special magic here"? If we can avoid redundant stuff we should do it.

Comment (jastice) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: I could imagine truncating common substrings, but that complexity would only be worth it if it's common enough. If it's easy though, I'd suggest only adding the hash if truncating would otherwise lead to name duplications, to avoid cryptic looking module names.

Comment (steveniemitz) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: that seems more complicated than worth it for this kind of edge case.  If you feel like adding that in later feel free to, but it's out of the scope of the amount of effort I'm willing to put into this fix.

Comment (unkarjedy) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala]: It's already better then it was when It didn't work at all

**Review by jastice — APPROVED** [review]

**Comment by jastice** [comment]
Thanks! I'll merge it now, unless you were going to add something, @steveniemitz ?

**Comment by steveniemitz** [comment]
Nope good on my side, thank you!

**Comment by jastice** [comment]
also cherrypicked to `idea222.x`: 2d84da

**Comment by steveniemitz** [comment]
thank you!

**Comment by steveniemitz** [comment]
I noticed this (and #616) wasn't in the 2022.1.16 release, do I need to submit a PR to add it to the idea221.release branch too?

**Comment by jastice** [comment]
If a fix is important to you I can also cherrypick it on 221 :)

**Comment by steveniemitz** [comment]
> If a fix is important to you I can also cherrypick it on 221 :)

That'd be amazing, thank you :D

**Pull request merged**
